### PR TITLE
Use int64 instead of long long in x86 intrinsics

### PR DESCRIPTION
**Pull request by alexbatashev**
According to C99 standard long long is at least 64 bit in size. However,
OpenCL C defines long long as 128 bit signed integer. This prevents
one to use x86 intrinsics when compiling OpenCL C code for x86 targets.
The patch changes long long to int64 which is guaranteed to be 64 bits
in size on all platforms.

### CODE REVIEW

**Comment by alexbatashev** [comment]
The patch itself seems to be working for me. But there's a problem with tests. The types become incompatible. I don't know why it happens and how to resolve these issues.

**Comment by alexbatashev** [comment]
Failing Tests (25):
    Clang :: CodeGen/builtins-x86.c
    Clang :: CodeGen/mmx-inline-asm.c
    Clang :: CodeGen/mmx-shift-with-immediate.c
    Clang :: CodeGen/pclmul-builtins.c
    Clang :: CodeGen/regcall.c
    Clang :: CodeGen/sse-builtins-dbg.c
    Clang :: CodeGen/sse-builtins.c
    Clang :: CodeGen/sse.c
    Clang :: CodeGen/vector.c
    Clang :: CodeGen/x86-GCC-inline-asm-Y-constraints.c
    Clang :: CodeGen/x86-builtins-vector-width.c
    Clang :: CodeGen/x86-crc-builtins.c
    Clang :: CodeGenCXX/mangle-ms-vector-types.cpp
    Clang :: Headers/c89.c
    Clang :: Headers/mm3dnow.c
    Clang :: Headers/ms-intrin.cpp
    Clang :: Headers/pmmintrin.c
    Clang :: Headers/wmmintrin.c
    Clang :: Headers/x86-intrinsics-headers-clean.cpp
    Clang :: Headers/xmmintrin.c
    Clang :: Modules/compiler_builtins.m
    Clang :: Sema/2010-05-31-palignr.c
    Clang :: Sema/implicit-ms-builtin-decl.c
    Clang :: Sema/warn-shadow-intrinsics.c
    Clang :: Sema/x86-builtin-palignr.c

File: clang/lib/Headers/immintrin.h [hunk 1]
```diff
@@ -10,6 +10,10 @@
 #ifndef __IMMINTRIN_H
 #define __IMMINTRIN_H
 
+#include <stdint.h>
+typedef int64_t __int64_t;
```
Comment (alexbatashev) [inline clang/lib/Headers/immintrin.h]: Temporary hacks until I find out portable way to define these types

**Comment by alexbatashev** [comment]
Failing Tests (3):
    Clang :: Headers/x86-intrinsics-headers-clean.cpp
    Clang :: Modules/compiler_builtins.m
    Clang :: Sema/implicit-ms-builtin-decl.c

**Comment by topperc** [comment]
Ideally I'd like to use long long when we're not using opencl. A large portion of these builtins are shared with gcc which uses long long. I'm not sure if that could cause surprises for users.

**Comment by alexbatashev** [comment]
@topperc thank you for your reply. I’ll take it into account.

File: clang/include/clang/Basic/BuiltinsX86_64.def [hunk 2]
```diff
@@ -21,87 +21,87 @@
 #  define TARGET_HEADER_BUILTIN(ID, TYPE, ATTRS, HEADER, LANG, FEATURE) BUILTIN(ID, TYPE, ATTRS)
 #endif
 
-TARGET_HEADER_BUILTIN(_BitScanForward64, "UcUNi*ULLi", "nh", "intrin.h", ALL_MS_LANGUAGES, "")
-TARGET_HEADER_BUILTIN(_BitScanReverse64, "UcUNi*ULLi", "nh", "intrin.h", ALL_MS_LANGUAGES, "")
+TARGET_HEADER_BUILTIN(_BitScanForward64, "UcUNi*UOi", "nh", "intrin.h", ALL_MS_LANGUAGES, "")
```
Comment (topperc) [inline clang/include/clang/Basic/BuiltinsX86_64.def]: Everything marked with ALL_MS_LANGUAGES are only available in MSVC compatbility mode. Do they need to be touched for OpenCL?

Comment (alexbatashev) [reply clang/include/clang/Basic/BuiltinsX86_64.def]: @topperc I didn’t know that. Thank you for pointing out.

**Comment by topperc** [comment]
What is the plan for the header files? If I'm reading this latest state correctly. Only the builtins have been changed? You still need to replace all long long references in headers with some sort of typedef or macro that's defined differently when OpenCL is enabled. Also need to change the definition of __m128i/__m256i/__m512i to not use long long in OpenCL.

**Comment by alexbatashev** [comment]
@topperc I am considering 2 options:
1. Do not touch headers. Just copy them to our source base and modify as we need. Pros: we don’t make anyone else suffer from this patch. Cons: we’ll be responsible to keeping it up to date. 
2. Define a new type that is long for OCL and long long for others. Pros: zero effort to maintain for us. Cons: community may not like this.

**Pull request closed without merging**